tools/buildFactory.ts

```diff
@@ -0,0 +1,230 @@
+import gulp from 'gulp';
+import tap from 'gulp-tap';
+import filter from 'gulp-filter';
+import merge from 'merge-stream';
+import ts from 'gulp-typescript';
+import terser from 'gulp-terser';
+import GulpUglify from 'gulp-uglify';
+import babel from 'gulp-babel';
+import rename from 'gulp-rename';
+import { SrcOptions } from 'vinyl-fs';
+import sourcemaps from 'gulp-sourcemaps';
+import tsMacroTransformer, { macros } from 'ts-macros';
+import pathsTransformer from 'ts-transform-paths';
+import rollupTypescript from 'rollup-plugin-typescript2';
+import { babel as rollupBabel } from '@rollup/plugin-babel';
+import { CustomTransformerFactory, Program } from 'typescript';
+import minifyPrivatesTransformer from 'ts-transformer-minify-privates';
+import tsTreeshaker, { DependencySolver } from './tsTreeshaker';
+import deleteEmpty from 'delete-empty';
+import ttypescript from 'ttypescript';
+
+/* eslint-disable @typescript-eslint/no-var-requires */
+const clean = require('gulp-clean') as () => NodeJS.ReadWriteStream;
+const rollup = require('gulp-rollup-2').rollup as (...params: unknown[]) => NodeJS.ReadWriteStream;
+/* eslint-enable @typescript-eslint/no-var-requires */
+
+function createProject(overrideSettings?: ts.Settings) {
+  return ts.createProject('tsconfig.json', {
+    ...overrideSettings,
+    getCustomTransformers: (program?: Program) => {
+      if (program === undefined) throw new Error('Program is undefined');
+
+      const pathsTransform = pathsTransformer(program);
+
+      const overrideGetCustomTransformers = overrideSettings?.getCustomTransformers;
+      if (typeof (overrideGetCustomTransformers) === 'string') {
+        throw new Error('getCustomTransformers should be a function');
+      }
+      const customTransformers = overrideGetCustomTransformers?.(program);
+
+      return {
+        before: [
+          tsMacroTransformer(program) as unknown as CustomTransformerFactory,
```

Comment:
Why need type conversion?

Reply:
The type of tsMacroTransformer is wrong, so I can't help it.